File: backend/internal/http/handlers/public.go
package handlers

import (
	"strings"

	"dw0rdwk/backend/internal/models"
	"dw0rdwk/backend/internal/repository"
	"dw0rdwk/backend/internal/service"

	"github.com/gofiber/fiber/v2"
)

type publicOrderSearchPayload struct {
	Account string `json:"account"`
}

type publicOrderRefreshPayload struct {
	Account string `json:"account"`
}

type publicOrderRow struct {
	ID              uint   `json:"id"`
	Platform        string `json:"platform"`
	School          string `json:"school"`
	StudentName     string `json:"studentName"`
	Account         string `json:"account"`
	CourseName      string `json:"courseName"`
	Status          string `json:"status"`
	DockingStatus   string `json:"dockingStatus"`
	Progress        string `json:"progress"`
	Remarks         string `json:"remarks"`
	Score           string `json:"score"`
	DurationMinutes int    `json:"durationMinutes"`
	CreatedAt       string `json:"createdAt"`
}

type publicOrderSearchResult struct {
	Account string           `json:"account"`
	Items   []publicOrderRow `json:"items"`
	Total   int              `json:"total"`
	Notice  string           `json:"notice"`
}

func PublicOrderSearch(repo *repository.OrderRepository, settings *repository.SettingRepository) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var req publicOrderSearchPayload
		if err := c.BodyParser(&req); err != nil {
			return fiber.NewError(fiber.StatusBadRequest, "invalid request body")
		}
		account := strings.TrimSpace(req.Account)
		if account == "" {
			return fiber.NewError(fiber.StatusBadRequest, "account is required")
		}
		orders, err := repo.ListPublicByAccount(c.UserContext(), account, 50)
		if err != nil {
			return err
		}
		values, _ := settings.All(c.UserContext())
		return OK(c, publicOrderSearchResult{
			Account: account,
			Items:   publicOrderRowsFromOrders(orders),
			Total:   len(orders),
			Notice:  strings.TrimSpace(values["open_query_notice"]),
		})
	}
}

func PublicOrderRefresh(repo *repository.OrderRepository, orderService *service.OrderService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		id, err := pathID(c)
		if err != nil {
			return err
		}
		var req publicOrderRefreshPayload
		if err := c.BodyParser(&req); err != nil {
			return fiber.NewError(fiber.StatusBadRequest, "invalid request body")
		}
		account := strings.TrimSpace(req.Account)
		if account == "" {
			return fiber.NewError(fiber.StatusBadRequest, "account is required")
		}
		order, err := repo.Find(c.UserContext(), id)
		if err != nil {
			return err
		}
		if order.Account != account {
			return fiber.NewError(fiber.StatusForbidden, "order does not belong to this account")
		}
		if err := orderService.MarkRefreshRequested(c.UserContext(), id); err != nil {
			return err
		}
		return OK(c, fiber.Map{"id": id})
	}
}

func publicOrderRowsFromOrders(orders []models.Order) []publicOrderRow {
	items := make([]publicOrderRow, 0, len(orders))
	for _, order := range orders {
		items = append(items, publicOrderRow{
			ID:              order.ID,
			Platform:        order.Platform,
			School:          order.School,
			StudentName:     order.StudentName,
			Account:         order.Account,
			CourseName:      order.CourseName,
			Status:          order.Status,
			DockingStatus:   order.DockingStatus,
			Progress:        order.Progress,
			Remarks:         order.Remarks,
			Score:           order.Score,
			DurationMinutes: order.DurationMinutes,
			CreatedAt:       order.CreatedAt.Format("2006-01-02 15:04:05"),
		})
	}
	return items
}

File: frontend/src/api/public.ts
import { getData, patchData, postData } from './http';

export interface PublicOrderRow {
  id: number;
  platform: string;
  school: string;
  studentName: string;
  account: string;
  courseName: string;
  status: string;
  dockingStatus: string;
  progress: string;
  remarks: string;
  score: string;
  durationMinutes: number;
  createdAt: string;
}

export interface PublicOrderSearchResult {
  account: string;
  items: PublicOrderRow[];
  total: number;
  notice: string;
}

export interface OrderEventRow {
  id: number;
  orderId: number;
  userId: number;
  level: string;
  source: string;
  eventType: string;
  content: string;
  progress: string;
  visibleToUser: boolean;
  createdAt: string;
}

export function searchPublicOrders(account: string) {
  return postData<PublicOrderSearchResult>('/api/v1/public/orders/search', { account });
}

export function refreshPublicOrder(id: number, account: string) {
  return postData<{ id: number }>(`/api/v1/public/orders/${id}/refresh`, { account });
}

export function fetchPublicOrderEvents(id: number, account: string) {
  return getData<{ items: OrderEventRow[] }>(`/api/v1/public/orders/${id}/events`, { account });
}

export function updatePublicOrderPassword(id: number, account: string, password: string) {
  return patchData<{ id: number }>(`/api/v1/public/orders/${id}/password`, { account, password });
}

export function resubmitPublicOrder(id: number, account: string) {
  return postData<{ id: number }>(`/api/v1/public/orders/${id}/resubmit`, { account });
}
